File: matlab/optimize_W.m
function W = optimize_W(X, W, S, lambda1, lambda2)

[n, b] = size(X);
D = 0.5 * (diag(sum(S,1)) + diag(sum(S,2)));
L = D - 0.5 * (S + S');
Q = zeros(b,b);
for i = 1:b
    Q(i,i) = 1/norm(W(i,:));
end
A = -(X' * X) + lambda1 * Q + lambda2 * (X' * L * X);

ww = rand(b, 1);
for i = 1:10
    m1 = A * ww;
    q = m1 / norm(m1);
    ww = q;
end
lambda1_A = abs(ww' * A * ww);

t = 0;
t_max = 15;
A_til = lambda1_A.*eye(b) - A;
while t < t_max
    M = A_til * W;
    [U, ~, V] = svd(M, 'econ');
    W = U * V';
    t=t+1;
end

File: python/optimize_W.py
# -*- coding: utf-8 -*-

import numpy as np
from numpy.random import rand

def optimize_W(X, W, S, lambda1, lambda2, s):

    b = X.shape[1]
    D = 0.5 * np.diag(np.sum(S, axis=1) + np.sum(S, axis=0))
    L = D - 0.5 * (S + S.T)
    row, col = np.diag_indices(b)
    Q = np.zeros([b,b])
    Q[row,col] = 1/np.linalg.norm(W, axis=1)
    A = -(X.T @ X) + lambda1 * Q + lambda2 * (X.T @ L @ X)
    #A = lambda1 * Q + lambda2 * (X.T @ L @ X)
    if s==0:
        lambda1_A = np.max(np.abs(np.linalg.eig(A)[0]))
    elif s==1:
        ww = rand(b,1)
        for i in range(20):
            m1 = A @ ww
            q = m1 / np.linalg.norm(m1)
            ww = q
        lambda1_A = np.abs(ww.T @ A @ ww)
    else:
        print('Warning: error input!!!')
        return;
    
    t = 0
    t_max = 15            
    A_til = lambda1_A * np.eye(b) - A
    T = np.zeros(t_max)
    while t < t_max:
        M = A_til @ W
        U, Sigma, VT = np.linalg.svd(M, full_matrices=0)
        W = U @ VT
        T[t] = np.trace(W.T @ A @ W)
        #print("T[t]: ", T[t])
        t = t + 1
    
    return W

    
